tools/discord/bot.py

```diff
@@ -0,0 +1,255 @@
+
+
+async def respond_fail_error(
+    interaction: discord.Interaction, error: str, tip: Optional[str] = None
+):
+    embed = discord.Embed(color=0xFF595E, title="Fail!", description=f"### {error}")
+    embed.set_footer(text=tip)
+
+    if (
+        interaction.response.type
+        == discord.InteractionResponseType.deferred_channel_message
+    ):
+        await interaction.followup.send(embed=embed, ephemeral=True)
+    else:
+        await interaction.response.send_message(embed=embed, ephemeral=True)
+
+
+async def respond_bug_error(interaction: discord.Interaction, error: str):
+    embed = discord.Embed(color=0xFFCA3A, title="BUG!", description=f"### {error}")
+    embed.set_footer(
+        text="This should never be visible. "
+        "Please report this to a Kinoko developer if you see this!"
+    )
+
+    if (
+        interaction.response.type
+        == discord.InteractionResponseType.deferred_channel_message
+    ):
+        await interaction.followup.send(embed=embed, ephemeral=True)
+    else:
+        await interaction.response.send_message(embed=embed, ephemeral=True)
+
+
+async def respond_krkg_success(interaction: discord.Interaction, file: discord.File):
+    if (
+        interaction.response.type
+        == discord.InteractionResponseType.deferred_channel_message
+    ):
+        await interaction.followup.send(file=file, ephemeral=True)
+    else:
+        await interaction.response.send_message(file=file, ephemeral=True)
+
+
+# ================================
+#     EVENTS
+# ================================
+
+
+@client.event
+async def on_ready():
+    await tree.sync()
+    print(f"Synced command tree and logged in as {client.user}")
+
+
+# ================================
+#     COMMANDS
+# ================================
+
+
+@tree.command(name="ping")
+async def command_ping(interaction: discord.Interaction):
+    await interaction.response.send_message("Pong!", ephemeral=True)
+
+
+@tree.command(name="generate_krkg")
+async def command_generate_krkg(
+    interaction: discord.Interaction, ghost: discord.Attachment
+):
+    global IS_GENERATING_KRKG
+
+    # Log when people generate KRKGs
+    print(f"{interaction.user.name} is requesting to generate a KRKG!")
+
+    # Check if the file claims to be a ghost
+    file_extension = ghost.filename.split(".")[-1]
+    if file_extension != "rkg":
+        await respond_generic_error(
+            interaction,
+            "File is not an RKG",
+            "Double check that the file extension is .rkg!",
+        )
+        return
+
+    # Check if the size is too small or big to be a ghost
+    if ghost.size < 0x8C or ghost.size > 0x2800:
+        await respond_generic_error(
+            interaction,
+            f"File is too {"small" if ghost.size < 0x8c else "big"} to be an RKG",
+        )
+        return
+
+    # We can only generate one KRKG at a time, due to how we I/O with Dolphin
+    if IS_GENERATING_KRKG:
```

Comment:
I'm a bit confused on why this global is necessary. Does the discord package spawn a new thread every time a command is triggered? If no, then this is unnecessary because there is only ever one "instance" of KRKG generation happening.

If yes, then you have a race condition here I think? If another thread hits this if-statement at the same time as another thread, then it may be possible for neither instance to enter this if-statement. Unless there's something I'm missing that's Python-specific

Reply:
You can do
```
from threading import Lock

LOCK_IS_GENERATING_KRKG = Lock()
IS_GENERATING_KRKG = False

LOCK_IS_GENERATING_KRKG.acquire()
if (IS_GENERATING_KRKG):
    LOCK_IS_GENERATING_KRKG.release()
    await respond_generic_error(
        interaction,
        "Already generating a KRKG",
        "I can only handle one ghost at a time. Try again later!",
    )
    return

IS_GENERATING_KRKG = True
LOCK_IS_GENERATING_KRKG.release()
```